### PR TITLE
feat(chapter1): add go code to chapter1

### PR DESCRIPTION
**Pull request by Reanon**
1.add go code to chapter_computational_complexity

### CODE REVIEW

File: codes/go/chapter_computational_complexity/leetcode_two_sum_test.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package chapter_computational_complexity
+
+import (
+	"testing"
+)
+
+func TestTwoSum(t *testing.T) {
```
Comment (krahets) [inline codes/go/chapter_computational_complexity/leetcode_two_sum_test.go:7]: Sorry, I'm not familiar with Go.
Can we merge the two files, `leetcode_two_sum_test.go` and `leetcode_two_sum.go` ?

File: codes/go/go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module github.com/reanon/hello-algo/codes/go
```
Comment (krahets) [inline codes/go/go.mod:1]: Is this file necessary?
It looks like pointing out the local environment.

File: codes/go/chapter_computational_complexity/leetcode_two_sum.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package chapter_computational_complexity
+
+// twoSumBruteForce
```
Comment (krahets) [inline codes/go/chapter_computational_complexity/leetcode_two_sum.go:3]: Could you comment on this function following the Java codes?

File: codes/go/chapter_computational_complexity/leetcode_two_sum.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+		for j := i + 1; i < size; j++ {
+			if nums[i]+nums[j] == target {
+				return []int{i, j}
+			}
+		}
+	}
+	return nil
+}
+
+// twoSumHashTable
```
Comment (krahets) [inline codes/go/chapter_computational_complexity/leetcode_two_sum.go:16]: Same as above.

File: docs/chapter_computational_complexity/space_time_tradeoff.md [hunk 5]
```diff
@@ -69,11 +87,26 @@
...
 === "Python"
 
     ```python title="leetcode_two_sum.py"
     
     ```
+
+=== "Go"
+
+    ```go title="leetcode_two_sum.go"
+    func twoSumHashTable(nums []int, target int) []int {
```
Comment (krahets) [inline docs/chapter_computational_complexity/space_time_tradeoff.md:102]: We prefer `space` to `tab` (equivalent to 4 spaces) to avoid strange bugs.

File: codes/go/chapter_computational_complexity/leetcode_two_sum.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package chapter_computational_complexity
+
+// twoSumBruteForce
+func twoSumBruteForce(nums []int, target int) []int {
+	size := len(nums)
+	for i := 0; i < size-1; i++ {
+		for j := i + 1; i < size; j++ {
+			if nums[i]+nums[j] == target {
+				return []int{i, j}
```
Comment (krahets) [inline codes/go/chapter_computational_complexity/leetcode_two_sum.go:9]: We prefer `space` to `tab` (equivalent to 4 spaces) to avoid strange bugs.

File: codes/go/chapter_computational_complexity/leetcode_two_sum.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+	}
+	return nil
+}
+
+// twoSumHashTable
+func twoSumHashTable(nums []int, target int) []int {
+	hashTable := map[int]int{}
+	for idx, val := range nums {
+		if preIdx, ok := hashTable[target-val]; ok {
+			return []int{preIdx, idx}
```
Comment (krahets) [inline codes/go/chapter_computational_complexity/leetcode_two_sum.go:21]: We prefer `space` to `tab` (equivalent to 4 spaces) to avoid strange bugs.

**Review by krahets — COMMENTED** [review]
Thank @Reanon for the translation!
There are still some issues needed to be clarified.

File: codes/go/chapter_computational_complexity/leetcode_two_sum_test.go (hunk 1 above)
Comment (Reanon) [reply codes/go/chapter_computational_complexity/leetcode_two_sum_test.go:7]: Go has a built-in testing command called `go test` and a package `testing` which combine to give a minimal but complete testing experience.So the two files are usually separated.

File: docs/chapter_computational_complexity/space_time_tradeoff.md (hunk 5 above)
Comment (Reanon) [reply docs/chapter_computational_complexity/space_time_tradeoff.md:102]: `tab` will be replaced with 4 spaces in the next patch

File: codes/go/go.mod (hunk 2 above)
Comment (Reanon) [reply codes/go/go.mod:1]: Go application project recommend module named as `github.com/YOUR-USER-OR-ORG-NAME/YOUR-REPO-NAME`. It will be replaced with  github.com/krahets/hello-algo/codes/go in the next patch

File: codes/go/chapter_computational_complexity/leetcode_two_sum.go (hunk 6 above)
Comment (Reanon) [reply codes/go/chapter_computational_complexity/leetcode_two_sum.go:9]: same as above

File: codes/go/chapter_computational_complexity/leetcode_two_sum.go (hunk 7 above)
Comment (Reanon) [reply codes/go/chapter_computational_complexity/leetcode_two_sum.go:21]: same as above

File: codes/go/chapter_computational_complexity/leetcode_two_sum_test.go (hunk 1 above)
Comment (krahets) [reply codes/go/chapter_computational_complexity/leetcode_two_sum_test.go:7]: Thanks for the clear explanation.

File: codes/go/go.mod (hunk 2 above)
Comment (krahets) [reply codes/go/go.mod:1]: I see. By the way, I will add a file header for each code file to reveal the file name, created time, and author.

For instance,

```java
/*
 * File: leetcode_two_sum.java
 * Created Time: 2022-11-25
 * Author: Krahets (krahets@163.com)
 */
```

You shall replace the content with yours in the consequent commits, thanks!

File: docs/chapter_computational_complexity/space_time_tradeoff.md (hunk 5 above)
Comment (Reanon) [reply docs/chapter_computational_complexity/space_time_tradeoff.md:102]: In Golang, Tabs for indentation is the official recommendation, Docker and Kubernets also use tabs. The correctness of the code is guaranteed by gofmt. More in https://go.dev/doc/effective_go#formatting

Comment (Reanon) [reply docs/chapter_computational_complexity/space_time_tradeoff.md:102]: > `tab` will be replaced with 4 spaces in the next patch

This comment is deprecated.

File: codes/go/go.mod (hunk 2 above)
Comment (Reanon) [reply codes/go/go.mod:1]: I get it, file header will be append like this in golang: 
```
// File: leetcode_two_sum.go
// Created Time: 2022-11-25
// Author: Krahets (krahets@163.com)
```

File: docs/chapter_computational_complexity/space_time_tradeoff.md (hunk 5 above)
Comment (krahets) [reply docs/chapter_computational_complexity/space_time_tradeoff.md:102]: Thanks for the clear description.

File: codes/go/go.mod (hunk 2 above)
Comment (krahets) [reply codes/go/go.mod]: Thanks @Reanon

**Review by krahets — APPROVED** [review]
Thanks for the work

**Comment by krahets** [comment]
@Reanon By the way, it is not recommended to use `force push` because it will overrides the remote master branch.

**Comment by krahets** [comment]
This PR is related to the issue https://github.com/krahets/hello-algo/issues/14 .

**Comment by krahets** [comment]
https://github.com/krahets/hello-algo/issues/16

**Comment by Reanon** [comment]
> 

`Git' s push --force` is to keep only one commit for clear. GitHub doesn't have the functionality of patches, maybe mutil-commits are better choice. The next PR will not use `push --force`.
